substrate/client/authority-discovery/src/worker/addr_cache.rs

```diff
@@ -172,8 +240,224 @@ fn addresses_to_peer_ids(addresses: &HashSet<Multiaddr>) -> HashSet<PeerId> {
+	}
+}
+
+/// Writes the content to a file at the specified path.
+fn write_to_file(path: &PathBuf, content: &str) -> io::Result<()> {
+	let mut file = File::create(path)?;
+	file.write_all(content.as_bytes())?;
+	file.flush()
+}
+
+/// Reads content from a file at the specified path and tries to JSON deserializes
+/// it into the specified type.
+fn load_from_file<T: DeserializeOwned>(path: &PathBuf) -> io::Result<T> {
+	let file = File::open(path)?;
+	let reader = BufReader::new(file);
+
+	serde_json::from_reader(reader).map_err(|e| {
+		error!(target: super::LOG_TARGET, "Failed to load from file: {}", e);
+		io::Error::new(io::ErrorKind::InvalidData, e)
+	})
+}
+
+/// Asynchronously writes content to a file using a background thread.
+/// Multiple consecutive writes in quick succession (before the current write is completed)
+/// will be **throttled**, and only the last write will be performed.
+#[derive(Clone)]
+pub struct ThrottlingAsyncFileWriter {
+	/// Each request to write content will send a message to this sender.
+	///
+	/// N.B. this is not passed in as an argument, it is an implementation
+	/// detail.
+	sender: mpsc::UnboundedSender<String>,
+}
+
+impl ThrottlingAsyncFileWriter {
+	/// Creates a new `ThrottlingAsyncFileWriter` for the specified file path,
+	/// the label is used for logging purposes.
+	pub fn new(purpose_label: String, path: impl Into<PathBuf>) -> Self {
+		let path = Arc::new(path.into());
+		let (sender, mut receiver) = mpsc::unbounded();
+
+		let path_clone = Arc::clone(&path);
+		thread::spawn(move || {
+			let mut latest: Option<String>;
+
+			while let Some(msg) = block_on(receiver.next()) {
+				latest = Some(msg);
+				while let Ok(Some(msg)) = receiver.try_next() {
+					latest = Some(msg);
+				}
+
+				if let Some(ref content) = latest {
+					if let Err(err) = write_to_file(&path_clone, content) {
+						error!(target: super::LOG_TARGET, "Failed to write to file for {}, error: {}", purpose_label, err);
+					}
+				}
+			}
+		});
+
+		Self { sender }
+	}
+
+	/// Write content to the file asynchronously, subsequent calls in quick succession
+	/// will be throttled, and only the last write will be performed.
+	///
+	/// The content is written to the file specified by the path in the constructor.
+	pub fn write(&self, content: impl Into<String>) {
+		let _ = self.sender.unbounded_send(content.into());
+	}
+
+	/// Serialize the value and write it to the file, asynchronously and throttled.
+	///
+	/// This calls `write` after serializing the value to a pretty JSON string.
+	pub fn write_serde<T: Serialize>(
+		&self,
+		value: &T,
+	) -> std::result::Result<(), serde_json::Error> {
+		let json = serde_json::to_string_pretty(value)?;
+		self.write(json);
+		Ok(())
+	}
+}
+
+/// Load contents of persisted cache from file, if it exists, and is valid. Create a new one
+/// otherwise, and install a callback to persist it on change.
+pub(crate) fn create_addr_cache(persistence_path: PathBuf) -> AddrCache {
```

Comment:
maybe this would be cleaner with:

```rust
{
	
	let loaded_or_empty: SerializableAddrCache = load_from_file(&persistence_path)
		.map_err(|_|Error::EncodingDecodingAddrCache(format!("Failed to load AddrCache from file: {}", persistence_path.display())))?;

	let async_file_writer = ThrottlingAsyncFileWriter::new("Persisted-AddrCache".to_owned(), &persistence_path);
	
	AddrCache::from((loaded_or_empty, move |cache| {
		let serializable = SerializableAddrCache::from(cache);
		debug!(target: super::LOG_TARGET, "Persisting AddrCache to file: {}", persistence_path.display());
		async_file_writer
			.write_serde(&serializable)
			.map_err(|e| Box::new(e) as Box<dyn std::fmt::Debug>)
	});
}
```